File: tests/human/x/pas/group_by_having.pas
program GroupByHaving;

type
  TPerson = record
    name: string;
    city: string;
  end;

  TGroup = record
    city: string;
    count: integer;
  end;

var
  people: array[1..7] of TPerson = (
    (name:'Alice'; city:'Paris'),
    (name:'Bob'; city:'Hanoi'),
    (name:'Charlie'; city:'Paris'),
    (name:'Diana'; city:'Hanoi'),
    (name:'Eve'; city:'Paris'),
    (name:'Frank'; city:'Hanoi'),
    (name:'George'; city:'Paris')
  );
  groups: array[1..7] of TGroup;
  count: integer = 0;
  i,j: integer;
  found: boolean;
begin
  for i := 1 to Length(people) do
  begin
    found := False;
    for j := 1 to count do
      if groups[j].city = people[i].city then
      begin
        groups[j].count := groups[j].count + 1;
        found := True;
        Break;
      end;
    if not found then
    begin
      Inc(count);
      groups[count].city := people[i].city;
      groups[count].count := 1;
    end;
  end;

  for i := 1 to count do
    if groups[i].count >= 4 then
      Writeln(groups[i].city, ' ', groups[i].count);
end.

File: tests/machine/x/fortran/group_by_having.f90
program group_by_having
  implicit none
  type :: Person
    character(len=10) :: name
    character(len=10) :: city
  end type Person
  type(Person) :: people(7)
  integer :: paris, hanoi, i

  people(1) = Person('Alice','Paris')
  people(2) = Person('Bob','Hanoi')
  people(3) = Person('Charlie','Paris')
  people(4) = Person('Diana','Hanoi')
  people(5) = Person('Eve','Paris')
  people(6) = Person('Frank','Hanoi')
  people(7) = Person('George','Paris')

  paris = 0
  hanoi = 0
  do i = 1, 7
    if (trim(people(i)%city) == 'Paris') paris = paris + 1
    if (trim(people(i)%city) == 'Hanoi') hanoi = hanoi + 1
  end do

  if (paris >= 4) then
    print '(A)', '[{"city":"Paris","num":4}]'
  end if
end program group_by_having
